stackslib/src/net/download/nakamoto.rs

```diff
@@ -0,0 +1,4283 @@
+            let Some(wt) = self
+                .wanted_tenures
+                .get(self.wanted_tenures.len().saturating_sub(2))
+            else {
+                return HashMap::new();
+            };
+            wt.clone()
+        };
+
+        // Run the confirmed downloader state machine set, since we could already be processing the
+        // highest complete tenure download.  NOTE: due to the way that we call this method, we're
+        // guaranteed that if the `tenure_downloads` downloader set has any downloads at all, they
+        // will only be for the highest complete tenure (i.e. we only call this method if we've
+        // already downloaded all confirmed tenures), so there's no risk of clobberring any other
+        // in-flight requests.
+        let new_confirmed_blocks = if self.tenure_downloads.inflight() > 0 {
+            self.download_confirmed_tenures(network, 0)
+        } else {
+            HashMap::new()
+        };
+
+        // Only run unconfirmed downloaders if we're _not_ busy obtaining the highest confirmed
+        // tenure.  The behavior here ensures that we first obtain the highest complete tenure, and
+        // then poll for new unconfirmed tenure blocks.
+        let (new_unconfirmed_blocks, new_highest_confirmed_downloaders) =
+            if self.tenure_downloads.inflight() > 0 {
+                (HashMap::new(), HashMap::new())
+            } else {
+                Self::run_unconfirmed_downloaders(
+                    &mut self.unconfirmed_tenure_downloads,
+                    network,
+                    &mut self.neighbor_rpc,
+                    sortdb,
+                    &burnchain_tip,
+                    chainstate,
+                    &highest_wanted_tenure,
+                    &unconfirmed_tenure,
+                )
+            };
+
+        // schedule downloaders for the highest-confirmed tenure, if we generated any
+        self.tenure_downloads
+            .add_downloaders(new_highest_confirmed_downloaders.into_iter());
+
+        // coalesce blocks -- maps consensus hash to map of block id to block
+        let mut coalesced_blocks: HashMap<ConsensusHash, HashMap<StacksBlockId, NakamotoBlock>> =
+            HashMap::new();
+        for blocks in new_unconfirmed_blocks
+            .into_values()
+            .chain(new_confirmed_blocks.into_values())
+        {
+            for block in blocks.into_iter() {
+                let block_id = block.header.block_id();
+                if let Some(block_map) = coalesced_blocks.get_mut(&block.header.consensus_hash) {
+                    block_map.insert(block_id, block);
+                } else {
+                    let mut block_map = HashMap::new();
+                    let ch = block.header.consensus_hash.clone();
+                    block_map.insert(block_id, block);
+                    coalesced_blocks.insert(ch, block_map);
+                }
```

Comment:
```suggestion
                let ch = block.header.consensus_hash.clone();
                coalesced_blocks.entry(ch).or_insert_with(|| HashMap::new()).insert(block_id, block);
```

Reply:
Skipping this one.  I don't think the `Entry` API helps improve clarity at all (just the opposite -- I think it adds a needless layer of indirection by requiring the developer to interact with a barely-abstracted-over raw pointer for the inner element).